packages/flutter/lib/src/widgets/selection_arena.dart

```diff
@@ -0,0 +1,190 @@
+        notifyListeners();
+      }
+      return;
+    }
+    // First check if we've entirely passed any selected texts.
+    //
+    //  *
+    //       HELLO THERE x
+    //
+    Offset selectionStart = selectionStartT;
+    Offset selectionEnd = selectionEndT;
+    bool didUpdate = false;
+    for (var data in _selectionCandidates.values) {
+      Rect? bounds = data.bounds;
+      if (bounds == null) {
+        continue;
+      }
+      var textSelectionStart = Offset(math.max(selectionStart.dx, bounds.topLeft.dx), math.max(selectionStart.dy, bounds.topLeft.dy));
+      var textSelectionEnd = Offset(math.min(selectionEnd.dx, bounds.bottomRight.dx), math.max(selectionEnd.dy, bounds.bottom));
+
+      // if (textSelectionStart.dy > bounds.bottom) {
+      //   continue;
+      // }
+      TextPosition startTextPosition = data.paragraph.getPositionForOffset(data.paragraph.globalToLocal(selectionStart));
+      TextPosition endTextPosition = data.paragraph.getPositionForOffset(data.paragraph.globalToLocal(selectionEnd));
+      if (startTextPosition == endTextPosition) {
+        continue;
+      }
+      didUpdate = true;
+      _selectionStates[data.paragraph] = _SelectionState(startTextPosition, endTextPosition);
+    }
+    if (didUpdate) {
+      notifyListeners();
+    }
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _gestureRecognizers[VerticalDragGestureRecognizer] = GestureRecognizerFactoryWithHandlers<VerticalDragGestureRecognizer>(
+      () => VerticalDragGestureRecognizer(),
+      (VerticalDragGestureRecognizer instance) {
+        instance
+          ..onDown = _handleDragDown
+          ..onStart = _handleDragStart
+          ..onUpdate = _handleDragUpdate
+          ..onEnd = _handleDragEnd
+          ..onCancel = _handleDragCancel
+          ..dragStartBehavior = DragStartBehavior.down
+          ..supportedDevices = <PointerDeviceKind>{PointerDeviceKind.mouse};
+      },
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return RawGestureDetector(
+      key: _gestureDetectorKey,
+      gestures: _gestureRecognizers,
+      behavior: HitTestBehavior.translucent,
+      excludeFromSemantics: true,
+      child: RenderHackyTextSelectionWidget(this, widget.child),
+    );
+  }
+}
+
+class RenderHackyTextSelectionWidget extends SingleChildRenderObjectWidget {
+  RenderHackyTextSelectionWidget(this._state, Widget child) : super(child: child);
+
+  final _SelectionAreaState _state;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return HackyTextSelectionPainter(_state);
+  }
+
+}
+
+class HackyTextSelectionPainter extends RenderProxyBox {
+  HackyTextSelectionPainter(this._state) {
+    _state.addListener(markNeedsPaint);
+  }
+
+
+  _SelectionAreaState _state;
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    super.paint(context, offset);
+    for (var entry in _state._selectionStates.entries) {
```

Comment:
I wonder whether we need to worry about something been draw on top of the selection, we can also make this an overlay entry that either been part of the modalRouter overlayEntries or a CompositedTransformFollower.

Reply:
Yes, definitely took some short cuts with this one 😅